File: chat-backend/src/main/java/com/smallchat/backend/global/utils/JwtProvider.java
package com.smallchat.backend.global.utils;

import com.smallchat.backend.user.domain.model.vo.Nickname;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.UUID;

@Component
public class JwtProvider {
    private final byte[] key;

    public JwtProvider(@Value("${auth.key}") byte[] key) {
        this.key = key;
    }


    public Tokens createTokens(UUID id, Nickname nickname) {
        String at = createToken(new TokenPayload(TokenType.ACCESS_TOKEN, id, nickname));
        String rt = createToken(new TokenPayload(TokenType.REFRESH_TOKEN, id, nickname));
        return new Tokens(at, rt);
    }

    public String createToken(TokenPayload payload) {
        return Jwts.builder()
                .claim("type", payload.tokenType())
                .claim("userId", payload.userId())
                .claim("nickname", payload.nickname().getValue())
                .expiration(payload.tokenType().getExpDate())
                .signWith(Keys.hmacShaKeyFor(key))
                .compact();
    }

    public TokenPayload parseFromBearer(String authorization) {
        String accessToken = authorization.replace("Bearer ", "");
        return parseToken(accessToken);
    }

    public TokenPayload parseToken(String token) {
        Claims payload = Jwts
                .parser()
                .verifyWith(Keys.hmacShaKeyFor(key))
                .build()
                .parseSignedClaims(token)
                .getPayload();
        String tokenType = Optional.ofNullable(payload.get("type", String.class)).orElseThrow(RuntimeException::new);
        String userId = Optional.ofNullable(payload.get("userId", String.class)).orElse("");
        String nickname = Optional.ofNullable(payload.get("nickname", String.class)).orElse("");
        return new TokenPayload(TokenType.getTokenType(tokenType), UUID.fromString(userId), new Nickname(nickname));
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/global/utils/JwtProvider.kt
package com.smallchat.backend.global.utils

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

/**
 * jws sign key가 필요한 메서드 노출
 * createTokens
 * parseFromBearer
 */
@Component
class JwtProvider(
    @param:Value("\${auth.key}")
    private val rawKey: ByteArray
) {

    private val key = Keys.hmacShaKeyFor(rawKey)
    private val parser = Jwts.parser().verifyWith(key).build()

    fun createTokens(id: String, nickname: String): Tokens {
        val at = buildToken(AuthenticatedUser(id, nickname, TokenType.ACCESS_TOKEN))
        val rt = buildToken(AuthenticatedUser(id, nickname, TokenType.REFRESH_TOKEN))
        return Tokens(at, rt)
    }

    private fun buildToken(payload: AuthenticatedUser): String {
        val (userId, nickname, tokenType) = payload
        return Jwts.builder()
            .claim("type", tokenType)
            .claim("userId", userId)
            .claim("nickname", nickname)
            .expiration(tokenType.expDate)
            .signWith(key)
            .compact()
    }

    fun parseFromBearer(authHeader: String): AuthenticatedUser {
        val accessToken = authHeader.substringAfter("Bearer ").trim()
        return parseToken(accessToken)
    }

    fun parseToken(token: String): AuthenticatedUser {
        try {
            val payload = parser
                .parseSignedClaims(token)
                .getPayload()
            val userId = payload["userId"] as String
            val nickname = payload["nickname"] as String
            val type = payload["type"] as String
            return AuthenticatedUser(userId, nickname, enumValueOf<TokenType>(type))
        } catch (error: Error) {
            throw RuntimeException("parse token failed")
        }
    }
}
